modules/substrate/src/lib.rs

```diff
@@ -13,3 +13,186 @@
 
 // You should have received a copy of the GNU General Public License
 // along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Substrate Bridge Pallet
+//!
+//! This pallet is an on-chain light client for chains which have a notion of finality. It's main
+//! jobs are to verify that a header is valid and that is has been finalized correctly. In doing so
+//! it will form a sparse header-chain of finalized headers which can be trusted checkpoints for
+//! other applications.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+// Runtime-generated enums
+#![allow(clippy::large_enum_variant)]
+
+use crate::verifier::ChainVerifier;
+use bp_substrate::{AuthoritySet, ImportedHeader, ScheduledChange};
+use frame_support::{decl_error, decl_module, decl_storage, dispatch};
+use frame_system::ensure_signed;
+use sp_runtime::traits::Header as HeaderT;
+use sp_std::{marker::PhantomData, prelude::*};
+
+mod verifier;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+type Hash<T> = <T as HeaderT>::Hash;
+type Number<T> = <T as HeaderT>::Number;
+
+pub trait Trait: frame_system::Trait {}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as SubstrateBridge {
+		/// Best finalized header.
+		// Maybe make this a HeaderId?
+		BestFinalized: Option<T::Header>;
+		/// Headers which have been imported into the pallet.
+		// Maybe made a HeaderId?
+		ImportedHeaders: map hasher(identity) T::Hash => Option<ImportedHeader<T::Header>>;
+		/// The current Grandpa Authority set.
+		CurrentAuthoritySet: AuthoritySet;
+		/// The next scheduled authority set change.
+		NextScheduledChange: ScheduledChange<Number<T::Header>>;
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// This header has failed basic verification.
+		InvalidHeader,
+		/// This header has not been finalized.
+		UnfinalizedHeader,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		/// Import a signed Substrate header into the runtime.
+		///
+		/// Will check for finality proofs, and if available finalize any already imported
+		/// headers which are finalized by the newly imported header.
+		#[weight = 0]
+		pub fn import_signed_header(
```

Comment:
You have both `import()` and `finalize()` API in single call. This means that during period `P` when the header is generated, but justification is not (yet), your pallet won't be able to import blocks. So if this call will have some large weight, your pallet may be out of sync for a while, because it'll: (1) need to wait till the end of `P` and submit justification (2) sync all descendants that have appeared during `P`

Reply:
What I mean that it should be meaningful decision, because it may became bottleneck in the future

Reply:
No, the `Verifier::import_header` function is still supposed to import blocks without justifications if it they don't provide a finality proof. They're marked as unfinalized, but the pallet has made _some_ progress.

It's true that all the blocks imported during `P` won't be finalized until we get a justification at the end of `P`, but that's how Grandpa works anyways, so I don't think it's an issue.

Reply:
> No, the Verifier::import_header function is still supposed to import blocks without justifications if it they don't provide a finality proof. They're marked as unfinalized, but the pallet has made some progress.

You do not have separate `finalize()` call - that's what I mean.

> It's true that all the blocks imported during P won't be finalized until we get a justification at the end of P, but that's how Grandpa works anyways, so I don't think it's an issue.

Without providing separate `finalize()` call for already-imported headers AND allowing to import headers on top of header that requires justification, you'll stuck.

Reply:
We mark old headers as finalized when a block with a justification is built on top of them. You're right in that you can't build a chain on top of a header that's supposed to have a justification until you get that header. So in that sense the chain is stuck.

I'm confused as to how this would work otherwise. Let's say that we have previously imported scheduled change that's supposed to take effect at block N. When we get to block N we expect that this block will have a justification, otherwise it's not valid and nobody should be building on top of it. This is because they'll eventually be on a fork when the header with justification at N comes in.

Am I misunderstanding this?

Reply:
> I'm confused as to how this would work otherwise. Let's say that we have previously imported scheduled change that's supposed to take effect at block N. When we get to block N we expect that this block will have a justification, otherwise it's not valid and nobody should be building on top of it. This is because they'll eventually be on a fork when the header with justification at N comes in.

Blocks generation and finalization are completely separate in Substrate, so the fact that some blocks are finalized and some are not **is not affecting** block production. So let's only concentrate on finalization now. Let's also omit forks for now.

At some block `B` you see the `ScheduledChange(N, S2)` digest that signals change to `S2` in `N` headers. That'd mean 2 things: (1) starting from block `B+N+1` blocks will be finalized with new set `S2` and (2) current set (`S1`) must emit justification for block `B+N`. The justification is produced asynchronously (remember that block production + finalization are separate) - i.e. block production may broadcast **valid** blocks `B+N`, `B+N+1`, `B+N+2`, ... before set `S1` will generate justification for `B+N`. The most important thing is that you can't accept any justifications from `S2` (i.e. from blocks >= `B+N+1`) until you'll accept justification for `B+N` (generated by `S1`).

That's why if there's no API to finalize already-imported block (`B+N`) + you keep accepting descendants of this block, then finalization would stuck - you're unable to import justification for `B+N` and you can't accept justifications for any of its descendants.

Reply:
@svyatonik I ended up splitting the header import and finality. Can you check that part again to see if it's alright?

Reply:
You still accept all justifications - i.e. in the example above you'll accept justification for block `B+N+100` from `S2` without waiting for justification for `B+N` from `S1`.

Why this is important, especially for light client (at least that's my understanding - probably there are some other concerns): let's say single malicious block author makes a block `X` where he signals authorities set change to some `MaliciousSet` on next block `X+1`. Then block `X+1` and some of its descendants are created by the same block author and then the `MaliciousSet` just generates justification for, say, `X+3`. This makes chain `X->X+1->X+2->X+3` final and irrevertible. At the same time, all others block authors and old GRANDPA authorities are keeping to produce valid headers + justifications. So there's now competing finalized forks (keep in mind that light client only looks at the headers => it can't ask runtime which header is actually valid from runtime PoV).

So justification of `B+N` (or `X+1` in latter example) by previous set `S1` is treated as a confirmation that the set `S1` agrees to transfer its 'power-to-finalize-blocks' to the next set `S2`. Without this confirmation, you can't trust `S2` => can't accept any justifications for blocks > `B+N`.

Hope it makes sense :)

Reply:
Actually, I've found that you've added `requires_justification` field to the header, but you never check it. Probably some commits are missing?

Reply:
I was unsure about the strategy to use when checking `requires_justification`. I've followed your advice and am iterating through the ancestors of the header that we're trying to finalize and checking if there's any ancestor which `requires_justification`. If they do then we don't have the "best" justification so we need to reject it.

I've added a test `does_not_import_future_justification()` which I think capture the correct logic here. Let me know if it looks better now.